gulpfile.js

```diff
@@ -2,52 +2,296 @@
+    let allFiles = [];
+    staticBundles[fileType].forEach(bundle => {
+        bundle.files.forEach(bFile => {
+            if (bFile.endsWith(fileExt)) {
+                allFiles.push(bFile);
+            }
+        });
+    });
+    return allFiles;
+};
+
+const handleError = task => {
+    return err => {
+        log.warn(colors.bold(colors.red(`[ERROR] ${task}:`)), colors.red(err));
+    };
+};
+
+const bundleCssFiles = bundle => {
+    let bundleFilename = `css/BUNDLES/${bundle.name}.css`;
+    if (global.watching) {
+        log.info(`building: ${bundleFilename}`);
+    }
+    let cssFiles = bundle.files.map(fileName => {
+        if (!fileName.endsWith('.css')) {
+            return fileName.replace(/\.(less|scss)$/i, '.css');
+        }
+        return fileName;
+    });
+    return gulp.src(cssFiles, {base: buildDir, 'cwd': buildDir})
+        .pipe(concat(bundleFilename))
+        .pipe(gulp.dest(buildDir));
+};
+
+const bundleJsFiles = bundle => {
+    let bundleFilename = `js/BUNDLES/${bundle.name}.js`;
+    if (global.watching) {
+        log.info(`building: ${bundleFilename}`);
+    }
+    return gulp.src(bundle.files, {base: buildDir, cwd: buildDir})
+        .pipe(gulpif(!production, sourcemaps.init()))
+        .pipe(concat(bundleFilename))
+        .pipe(gulpif(!production, sourcemaps.write({
+            'includeContent': true
+        })))
+        .pipe(gulp.dest(buildDir));
+};
+
+/***********************
+ * Start Tasks
+ */
+
+/**
+ * Delete the static_build directory and start fresh.
+ */
+gulp.task('clean', () => {
+    return del([buildDir]);
+});
+
+/**
+ * Copy assets from various sources into the static_build dir for processing.
+ */
+gulp.task('assets', () => {
+    return gulp.src([
+        'media/**/*',
+        'node_modules/@mozilla-protocol/core/**/*',
+        '!node_modules/@mozilla-protocol/core/*'])
+        .pipe(gulpif(global.watching, cached('all', cachedOpts)))
+        .pipe(gulp.dest(buildDir));
+});
+
+/**
+ * Find all SASS files from bundles in the static_build directory and compile them.
+ */
+gulp.task('sass', ['assets'], () => {
+    return gulp.src(allBundleFiles('css', '.scss'), {base: buildDir, cwd: buildDir})
+        //filter out unchanged scss files, only works when watching
+        .pipe(gulpif(global.watching, cached('sass', cachedOpts)))
+        .pipe(gulpif(!production, sourcemaps.init()))
+        .pipe(sass({
+            sourceComments: !production,
+            outputStyle: production ? 'compressed' : 'nested'
+        }).on('error', handleError('SASS')))
+        .pipe(gulpif(!production, sourcemaps.write({
+            'includeContent': true
+        })))
+        .pipe(gulp.dest(buildDir));
+});
+
+/**
+ * Watch and only compile those SASS files that have changed
+ */
+gulp.task('sass:watch', () => {
+    return watch(buildDir + '/css/**/*.scss', {base: buildDir})
+        // filter out internal imports (files starting with "_" )
```

Comment:
From my local testing, this seems to have a side effect in that changes to imported scss files aren't currently picked up at all. The only way I can visibly see a change to an imported style sheet (e.g. `_sub_nav.scss`) is to also force a change in the parent stylesheet (e.g. `firefox/home.scss`). I know this is a somewhat tricky thing we've discessed earlier in this pull request, but this is going to be somewhat of a pain point for front-end dev :(

Reply:
Agreed. It's definitely a missing piece. I tried gulp-sass-inheritance but couldn't get it working. Only thing I can think for now is to add a watcher specifically for the `_*.scss` files that will rebuild all of the bundles. It's not very efficient, but it would at least not force a dev to quit everything and restart. To be clear all you'd need to do is restart everything since it does a full build at start, but that's a pain as you said.

Reply:
I would go a little stronger than "somewhat of a pain point". 😝 IMO, this is a blocker for front-end dev.

I think it's definitely worth trying a separate watcher for `.scss` files prefixed with `_`. We'll see what performance is like rebuilding all bundles that way.

Reply:
@alexgibson @jpetto I've just pushed an update that will rebuild all SASS bundles when a library file (one that starts with `_`) is changed. This isn't optimal, but does work and isn't too slow in mytesting. Let me know what you think.

Reply:
@alexgibson I've just rebased this and pushed my demo again. I think this is ready for final review.

Reply:
Agree this isn't optimal, but it does work and i don't think it should interfere with workflow to the extent this should be a blocker. I'd like to get the other front-end developers to kick the tyres of this PR though.